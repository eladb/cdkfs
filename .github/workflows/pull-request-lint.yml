# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@01d5fd8a8ebb9aafe902c40c53f0f4744f7381eb
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
