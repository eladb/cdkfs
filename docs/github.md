# GitHub

By default, many projects are initialized with `GitHub` component to enabled GitHub as the default provider for CI/CD workflows. See https://docs.github.com/en/actions for more information.

The use of GitHub (and generating corresponding files in `.github`) can be disabled by specifying `github: false` in your project options.

## Workflows

TODO

See the `GitHub`, `GithubWorkflow`, and `Job` types in the [API reference](./api/API.md) for currently available APIs.

Example code of creating a GitHub workflow: <https://github.com/projen/projen/blob/65b4194c163f47ba4842981b0c92dbe516be787b/src/github/auto-approve.ts#L67-L105>

### Stale workflow

A "stale" workflow can be added which will automatically close issues or pull
requests on your GitHub repository after time has passed without the issue
seeing any comments or updates. You can enable as shown below:

```ts
// or PythonProject, etc.
new typescript.TypeScriptProject({
  stale: true,
  staleOptions: {
    issues: {
      closeMessage: "closing pull request",
      staleLabel: "I-AM-STALE",
      daysBeforeStale: 180,
    },
  }
})
```

Check the API reference for a list of all available options.

When enabled, by default issues with no activity with will be marked as stale
after 60 days and closed within 7 days, and pull requests with no activity will
be marked as stale after 14 days and closed within 2 days.
