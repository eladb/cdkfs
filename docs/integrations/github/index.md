---
sidebar_position: 1
---

# GitHub Integration


This is just a placeholder doc. You can contribute to this project! Please see the [Contributing Guide](https://github.com/projen/projen/blob/main/docusaurus/README.md) for more information.
