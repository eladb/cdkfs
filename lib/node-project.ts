import { Project, ProjectOptions } from './project';
import { JsonFile } from './json';
import { Semver } from './semver';
import { IgnoreFile } from './ignore-file';
import { License } from './license';
import { GENERATION_DISCLAIMER, PROJEN_RC, PROJEN_VERSION } from './common';
import { Lazy } from 'constructs';
import { Version } from './version';
import { GithubWorkflow } from './github-workflow';

const ANTITAMPER_COMMAND = [
  {
    name: 'Anti-tamper check',
    run: 'git diff --exit-code',
  },
];

export interface CommonOptions {
  readonly bundledDependencies?: string[];
  readonly dependencies?: Record<string, Semver>;
  readonly devDependencies?: Record<string, Semver>;
  readonly peerDependencies?: Record<string, Semver>;
  readonly peerDependencyOptions?: PeerDependencyOptions;
  readonly bin?: Record<string, string>;
  readonly keywords?: string[];

  /**
   * Should we commit `package.json` to git or ignore?
   *
   * @default false By default `package.json` is *not* committed. This means
   * that after you check out a repository you must run `npx projen` to
   * bootstrap it.
   */
  readonly commitPackageJson?: boolean;

  /**
   * Version of projen to install.
   *
   * @default - latest version
   */
  readonly projenVersion?: Semver;

  /**
   * Indicates of "projen" should be installed as a devDependency.
   *
   * @default true
   */
  readonly projenDevDependency?: boolean;

  /**
   * Define a GitHub workflow for building PRs.
   * @default true
   */
  readonly buildWorkflow?: boolean;

  /**
   * Define a GitHub workflow for releasing from "master" when new versions are bumped.
   * @default true
   */
  readonly releaseWorkflow?: boolean;

  /**
   * Workflow steps to use in order to bootstrap this repo.
   *
   * @default - [ { run: `npx projen${PROJEN_VERSION}` }, { run: 'yarn install --frozen-lockfile' } ]
   */
  readonly workflowBootstrapSteps?: any[];

  /**
   * Container image to use for GitHub workflows.
   *
   * @default - default image
   */
  readonly workflowContainerImage?: string;

  /**
   * Automatically release to npm when new versions are introduced.
   * @default true
   */
  readonly releaseToNpm?: boolean;

  /**
   * Checks that after build there are no modified files onn git.
   * @default true
   */
  readonly antitamper?: boolean;
}

export interface NodeProjectOptions extends ProjectOptions, CommonOptions {
  readonly name: string;
  readonly description?: string;
  readonly repository?: string;
  readonly authorName?: string;
  readonly authorEmail?: string;
  readonly license?: string;
  readonly stability?: string;
  readonly gitignore?: string[];
  readonly npmignore?: string[];
}

export class NodeProject extends Project {
  public readonly npmignore: IgnoreFile;

  private readonly peerDependencies: Record<string, string> = { };
  private readonly devDependencies: Record<string, string> = { };
  private readonly dependencies: Record<string, string> = { };
  private readonly bundledDependencies: string[] = [];
  private readonly scripts: Record<string, string> = { };
  private readonly bin: Record<string, string> = { };

  private readonly manifest: any;
  private readonly testCommands = new Array<string>();
  private readonly _version: Version;

  /**
   * The PR build GitHub workflow. `undefined` if `buildWorkflow` is disabled.
   */
  protected readonly buildWorkflow?: NodeBuildWorkflow;

  /**
   * The release GitHub workflow. `undefined` if `releaseWorkflow` is disabled.
   */
  protected readonly releaseWorkflow?: NodeBuildWorkflow;

  constructor(options: NodeProjectOptions) {
    super(options);

    this.manifest = {
      '//': GENERATION_DISCLAIMER,
      name: options.name,
      description: options.description,
      main: 'lib/index.js',
      repository: !options.repository ? undefined : {
        type: 'git',
        url: options.repository,
      },
      bin: this.bin,
      scripts: this.scripts,
      author: {
        name: options.authorName,
        email: options.authorEmail,
      },
      devDependencies: this.devDependencies,
      peerDependencies: this.peerDependencies,
      dependencies: this.dependencies,
      bundledDependencies: this.bundledDependencies,
      keywords: options.keywords,
    };

    const commitPackageJson = options.commitPackageJson ?? false;
    new JsonFile(this, 'package.json', {
      committed: commitPackageJson,
      obj: this.manifest,
    });

    this.addDependencies(options.dependencies ?? {});
    this.addPeerDependencies(options.peerDependencies ?? {});
    this.addDevDependencies(options.devDependencies ?? {});
    this.addBundledDependencies(...options.bundledDependencies ?? []);

    this.npmignore = new IgnoreFile(this, '.npmignore');

    this.addDefaultGitIgnore();

    // set license and produce license file
    const license = options.license ?? 'Apache-2.0';
    this.manifest.license = license;
    new License(this, license);

    this.addScripts({ projen: `node ${PROJEN_RC} && yarn install` });
    this.addScripts({ 'projen:upgrade': 'yarn upgrade projen && yarn projen' });

    this.npmignore.comment('exclude project definition from npm module');
    this.npmignore.exclude(`/${PROJEN_RC}`);

    this.npmignore.comment('make sure to commit projen definition');
    this.gitignore.include(`/${PROJEN_RC}`);

    this.addBins(options.bin ?? { });

    const projen = options.projenDevDependency ?? true;
    if (projen) {
      const projenVersion = options.projenVersion ?? Semver.caret(PROJEN_VERSION);
      this.addDevDependencies({ projen: projenVersion });
    }

    this.addScripts({ test: Lazy.stringValue({ produce: () => this.renderTestCommand() }) });

    // version is read from a committed file called version.json which is how we bump
    this._version = new Version(this);
    this.manifest.version = this.version;

    if (options.buildWorkflow ?? true) {
      this.buildWorkflow = new NodeBuildWorkflow(this, 'Build', {
        trigger: { pull_request: { } },
        bootstrapSteps: options.workflowBootstrapSteps,
        image: options.workflowContainerImage,
        antitamper: options.antitamper,
      });
    }

    if (options.releaseWorkflow ?? true) {
      this.releaseWorkflow = new NodeBuildWorkflow(this, 'Release', {
        trigger: { push: { branches: [ 'master' ] } },
        uploadArtifact: true,
        bootstrapSteps: options.workflowBootstrapSteps,
        image: options.workflowContainerImage,
        antitamper: options.antitamper,
      });

      if (options.releaseToNpm) {
        this.releaseWorkflow.addJobs({
          release_npm: {
            'name': 'Release to NPM',
            'needs': this.releaseWorkflow.buildJobId,
            'runs-on': 'ubuntu-latest',
            'steps': [
              {
                'name': 'Download build artifacts',
                'uses': 'actions/download-artifact@v1',
                'with': {
                  'name': 'dist',
                },
              },
              {
                'name': 'Release',
                'run': 'npx -p jsii-release jsii-release-npm',
                'env': {
                  'NPM_TOKEN': '${{ secrets.NPM_TOKEN }}',
                },
              },
            ],
          },
        });
      }
    }
  }

  /**
   * Returns the current version of the project.
   */
  public get version() {
    return this._version.current;
  }

  public addBins(bins: Record<string, string>) {
    for (const [ k, v ] of Object.entries(bins)) {
      this.bin[k] = v;
    }
  }

  public addDependencies(deps: { [module: string]: Semver }, bundle = false) {
    for (const [ k, v ] of Object.entries(deps)) {
      this.dependencies[k] = v.spec;

      if (bundle) {
        this.addBundledDependencies(k);
      }
    }
  }

  public addBundledDependencies(...deps: string[]) {
    for (const dep of deps) {
      if (!(dep in this.dependencies)) {
        throw new Error(`unable to bundle "${dep}". it has to also be defined as a dependency`);
      }

      if (dep in this.peerDependencies) {
        throw new Error(`unable to bundle "${dep}". it cannot appear as a peer dependency`);
      }

      this.bundledDependencies.push(dep);
    }
  }

  public addDevDependencies(deps: { [module: string]: Semver }) {
    for (const [ k, v ] of Object.entries(deps)) {
      this.devDependencies[k] = v.spec;
    }
  }

  public addPeerDependencies(deps: { [module: string]: Semver }, options: PeerDependencyOptions = {}) {
    const pinnedDevDependency = options.pinnedDevDependency ?? true;
    for (const [ k, v ] of Object.entries(deps)) {
      this.peerDependencies[k] = v.spec;

      if (pinnedDevDependency) {
        this.addDevDependencies({ [k]: Semver.pinned(v.version) });
      }
    }
  }
  
  public addScripts(scripts: { [name: string]: string }) {
    for (const [ name, command ] of Object.entries(scripts)) {
      this.scripts[name] = command;
    }
  }

  public addTestCommands(...commands: string[]) {
    this.testCommands.push(...commands);
  }

  public addFields(fields: { [name: string]: any }) {
    for (const [ name, value ] of Object.entries(fields)) {
      this.manifest[name] = value;
    }
  }

  private renderTestCommand() {
    if (this.testCommands.length === 0) {
      return "echo 'no tests'";
    } else {
      return this.testCommands.join(' && ');
    }
  }

  private addDefaultGitIgnore()  {

    this.gitignore.exclude(
      '# Logs',
      'logs',
      '*.log',
      'npm-debug.log*',
      'yarn-debug.log*',
      'yarn-error.log*',
      'lerna-debug.log*',

      '# Diagnostic reports (https://nodejs.org/api/report.html)',
      'report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json',

      '# Runtime data',
      'pids',
      '*.pid',
      '*.seed',
      '*.pid.lock',

      '# Directory for instrumented libs generated by jscoverage/JSCover',
      'lib-cov',

      '# Coverage directory used by tools like istanbul',
      'coverage',
      '*.lcov',

      '# nyc test coverage',
      '.nyc_output',

      '# Compiled binary addons (https://nodejs.org/api/addons.html)',
      'build/Release',

      '# Dependency directories',
      'node_modules/',
      'jspm_packages/',

      '# TypeScript cache',
      '*.tsbuildinfo',


      '# Optional eslint cache',
      '.eslintcache',

      '# Output of \'npm pack\'',
      '*.tgz',

      '# Yarn Integrity file',
      '.yarn-integrity',

      '# parcel-bundler cache (https://parceljs.org/)',
      '.cache',
    );
  }
}

export interface PeerDependencyOptions {
  /**
   * Automatically add a pinned dev dependency.
   * @default true
   */
  readonly pinnedDevDependency?: boolean;
}

const DEFAULT_WORKFLOW_BOOTSTRAP = [
  { run: `npx projen@${PROJEN_VERSION}` },
  { run: 'yarn install --frozen-lockfile' },
];

export interface NodeBuildWorkflowOptions  {
  /**
   * @default - default image
   */
  readonly image?: string;

  /**
   * Workflow steps to use in order to bootstrap this repo.
   * @default - [ { run: `npx projen${PROJEN_VERSION}` }, { run: 'yarn install --frozen-lockfile' } ]
   */
  readonly bootstrapSteps?: any[];

  readonly uploadArtifact?: boolean;

  /**
   * Checks that after build there are no modified files onn git.
   * @default true
   */
  readonly antitamper?: boolean;

  readonly trigger: { [event: string]: any };
}

export class NodeBuildWorkflow extends GithubWorkflow {
  public readonly buildJobId: string;

  constructor(project: Project, name: string, options: NodeBuildWorkflowOptions) {
    super(project, name);

    this.buildJobId = 'build';

    this.on(options.trigger);

    const job: Record<string, any> = {
      'runs-on': 'ubuntu-latest',
      'steps': [
        { uses: 'actions/checkout@v2' },
        ...options.bootstrapSteps ?? DEFAULT_WORKFLOW_BOOTSTRAP,
        { run: 'yarn build' },
        ...(options.antitamper ?? true) ? ANTITAMPER_COMMAND : [],
      ],
    };

    if (options.image) {
      job.container = { image: options.image };
    }

    if (options.uploadArtifact) {
      job.steps.push({
        name: 'Upload artifact',
        uses: 'actions/upload-artifact@v1',
        with: {
          name: 'dist',
          path: 'dist',
        },
      });
    }

    this.addJobs({ [this.buildJobId]: job });
  }
}
