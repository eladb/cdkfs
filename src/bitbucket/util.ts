

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}