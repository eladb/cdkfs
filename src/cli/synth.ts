import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs-extra';
import { PROJEN_RC } from '../common';
import * as logging from '../logging';

const projenfile = path.resolve(PROJEN_RC);
const projen = path.join(__dirname, '..');

export function synth() {
  if (!fs.existsSync(projenfile)) {
    logging.error(`Unable to find ${projenfile}. Use "projen new" to create a new project.`);
    process.exit(1);
  }

  // if node_modules/projen is not a directory or does not exist, create a
  // temporary symlink to the projen that we are currently running in order to
  // allow .projenrc.js to `require()` it.
  logging.info('Synthesizing project...');
  const projenModulePath = path.resolve('node_modules', 'projen');
  if (!fs.existsSync(path.join(projenModulePath, 'package.json')) || !fs.statSync(projenModulePath).isDirectory()) {
    fs.removeSync(projenModulePath);
    fs.mkdirpSync('node_modules');
    fs.symlinkSync(projen, projenModulePath, (os.platform() === 'win32') ? 'junction' : null);
  }

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  require(projenfile);
}