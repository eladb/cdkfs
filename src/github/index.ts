export * from './github';

export * from './dependabot';
export * from './workflows';
export * from './mergify';
export * from './pr-template';
export * from './mergify';
export * from './auto-merge';
export * from './auto-approve';
export * from './stale';

export * from './task-workflow';

export * as workflows from './workflows-model';
