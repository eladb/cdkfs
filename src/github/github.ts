import { Component } from '../component';
import { FileBase } from '../file';
import { Project } from '../project';
import { Dependabot, DependabotOptions } from './dependabot';
import { Mergify } from './mergify';
import { PullRequestTemplate } from './pr-template';
import { GithubWorkflow } from './workflows';

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;
}

export class GitHub extends Component {
  /**
   * The `Mergify` configured on this repository. This is `undefined` if Mergify
   * was not enabled when creating the repository.
   */
  public readonly mergify?: Mergify;

  /**
   * The .gitattributes file for this repository.
   */
  private readonly gitattributes: GitattributesFile;

  public constructor(project: Project, options: GitHubOptions) {
    super(project);

    this.gitattributes = new GitattributesFile(project);

    // Contents  of the .projen/ directory are generated by projen
    this.annotateGenerated('/.projen/**');
    // The .gitattributes file itself is generated
    this.annotateGenerated(`/${this.gitattributes.path}`);

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this);
    }
  }

  public addWorkflow(name: string) {
    return new GithubWorkflow(this, name);
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }

  /**
   * Marks the provided file(s) as being generated. This is achieved using the
   * github-linguist attributes. Generated files do not count against the
   * repository statistics and language breakdown.
   *
   * @param glob the glob pattern to match (could be a file path).
   *
   * @see https://github.com/github/linguist/blob/master/docs/overrides.md
   */
  public annotateGenerated(glob: string): void {
    this.gitattributes.addAttributes(glob, 'linguist-generated');
  }
}

class GitattributesFile extends FileBase {
  private readonly attributes = new Map<string, Set<string>>();

  public constructor(project: Project) {
    super(project, '.gitattributes');
  }

  public addAttributes(glob: string, ...attributes: readonly string[]) {
    if (!this.attributes.has(glob)) {
      this.attributes.set(glob, new Set());
    }
    const set = this.attributes.get(glob)!;
    for (const attribute of attributes) {
      set.add(attribute);
    }
  }

  protected synthesizeContent(): string | undefined {
    // We can assume the file map is never empty.
    const entries = Array.from(this.attributes.entries())
      .sort(([l], [r]) => l.localeCompare(r));

    const maxLen = Math.max(...entries.map(([glob]) => glob.length));
    return [
      `# ${FileBase.PROJEN_MARKER}`,
      '',
      ...entries.map(([name, attributes]) => `${name.padEnd(maxLen, ' ')}\t${Array.from(attributes).join(' ')}`),
    ].join('\n');
  }
}
