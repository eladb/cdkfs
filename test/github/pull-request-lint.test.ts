import { PullRequestLint } from '../../src/github/pull-request-lint';
import { NodeProject, NodeProjectOptions } from '../../src/node-project';
import { synthSnapshot } from '../../src/util/synth';

test('default', () => {
  // GIVEN
  const project = createProject();

  // WHEN
  new PullRequestLint(project.github!);

  // THEN
  const snapshot = synthSnapshot(project);
  expect(snapshot['.github/workflows/pull-request-lint.yml']).toBeDefined();
  expect(snapshot['.github/workflows/pull-request-lint.yml']).toMatchSnapshot();
});

describe('semantic titles', () => {
  test('configure scopes', () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestLint(project.github!, {
      semanticTitle: true,
      semanticTitleOptions: {
        types: ['feat', 'fix'],
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/workflows/pull-request-lint.yml']).toMatchSnapshot();
  });

  test('require scope', () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestLint(project.github!, {
      semanticTitle: true,
      semanticTitleOptions: {
        requireScope: true,
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/workflows/pull-request-lint.yml']).toMatchSnapshot();
  });
});

type ProjectOptions = Omit<NodeProjectOptions, 'outdir' | 'defaultReleaseBranch' | 'name'>;
function createProject(options: ProjectOptions = {}): NodeProject {
  return new NodeProject({
    defaultReleaseBranch: 'main',
    name: 'node-project',
    githubOptions: {
      pullRequestLintOptions: {
        semanticTitle: false,
      },
    },
    ...options,
  });
}