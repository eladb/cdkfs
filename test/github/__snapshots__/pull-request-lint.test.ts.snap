// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v5.0.2
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
          githubBaseUrl: \${{ github.action_url }}
"
`;

exports[`semantic titles configure scopes 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v5.0.2
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
          requireScope: false
          githubBaseUrl: \${{ github.action_url }}
"
`;

exports[`semantic titles require scope 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v5.0.2
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: true
          githubBaseUrl: \${{ github.action_url }}
"
`;
